images/cilium/tests/main.tf

```diff
@@ -16,36 +24,38 @@ data "oci_exec_test" "operator-version" {
   digest = var.digests.operator
 }
 
-# TODO:
-# To run this test, we need to create a cluster that looks like this
-#
-# k3d cluster create -i cgr.dev/chainguard/k3s:latest \
-#     --k3s-arg '--disable=traefik@server:*' \
-#     --k3s-arg '--disable=metrics-server@server:*' \
-#     --k3s-arg '--disable-network-policy@server:*' \
-#     --k3s-arg '--flannel-backend=none@server:*' \
-#
-# for node in k3d-k3s-default-server-0; do
-#   docker exec -i $node /bin/sh <<-EOF
-#       mount bpffs -t bpf /sys/fs/bpf
-#       mount --make-shared /sys/fs/bpf
-#       mkdir -p /run/cilium/cgroupv2
-#       mount -t cgroup2 none /run/cilium/cgroupv2
-#       mount --make-shared /run/cilium/cgroupv2/
-#
-#
-# resource "helm_release" "cilium" {
-#   repository       = "https://helm.cilium.io"
-#   chart            = "cilium"
-#   name             = "cilium"
-#   namespace        = "kube-system"
-
-#   values = [jsonencode({
-#     operator = {
-#       image = {
-#         override = var.digests.operator
-#       }
-#       replicas = 1
-#     }
-#   })]
-# }
+data "oci_exec_test" "create-cluster" {
+  script = "${path.module}/create-cluster.sh"
+  digest = var.digests.operator
+}
+
+resource "helm_release" "cilium" {
+  depends_on = [data.oci_exec_test.create-cluster]
+  repository = "https://helm.cilium.io"
+  chart      = "cilium"
+  name       = "cilium"
+  namespace  = "kube-system"
+
+  values = [jsonencode({
+    operator = {
+      image = {
+        override = var.digests.operator
+      }
+      replicas = 1
+    }
+  })]
+}
+
+data "oci_exec_test" "wait-for-pods" {
+  depends_on = [helm_release.cilium]
+  # Due to lack of CNI, all pods in `kube-system` will be in `Pending` state.
+  # If they are 'Ready', then our installation is working.
+  script = "kubectl --context k3d-cilium-test wait --for condition=Ready -n kube-system pods --all --timeout=300s"
+  digest = var.digests.operator
+}
+
+data "oci_exec_test" "delete-cluster" {
```

Comment:
does this delete the cluster during failure cases? if not we might consider putting this all in a single `oci_exec_test` with a `trap`, since you're not relying on the `helm_release` readiness signal anyways.

my only concern is leaving dangling clusters during flakey tests. in CI it doesn't really matter, but can be a real pain locally.

Reply:
additionally, that might make it easier in the future to use the cilium cli to "properly" check readiness with `cilium status --wait`

Reply:
I like the idea of using `cilium` CLI to use their `cilium connectivity test` as well. However, in case of failed test on CI it is very hard to know what's going on without Pods log uploaded (post-submit flake is even harder).

Locally, this pain is felt more by the people working on the image itself and someone who may rebuild everything, but it is probably a good trade off at least in the beginning when we haven't flushed out all the flakes. I'd still move everything to a single script, but I'll keep the cluster in case of failures.

Reply:
Done